File: travelmate/src/services/authService.ts
//Authentication service - handles all auth-related API calls
import { User } from "../models";
import { httpClient } from "../utils/http-client";
import { logger } from "../utils/logger";
import { API_ENDPOINTS, buildAuthUrl } from "./api";

export const authService = {
  async verifyEmail(token: string) {
    try {
      logger.info("Verifying email", { token }, "AuthService");
      return await httpClient.verifyEmail(
        buildAuthUrl(API_ENDPOINTS.AUTH.VERIFY_EMAIL),
        token
      );
    } catch (error) {
      logger.error("Email verification failed", error, "AuthService");
      throw error;
    }
  },

  async resendVerification() {
    try {
      logger.info("Resending verification email", {}, "AuthService");
      return await httpClient.post(
        buildAuthUrl(API_ENDPOINTS.AUTH.RESEND_VERIFICATION),
        {},
        { skipAuth: false }
      );
    } catch (error) {
      logger.error("Resend verification failed", error, "AuthService");
      throw error;
    }
  },

  async getUserName(userIds: string[]) {
    try {
      logger.info("Fetching user names", { userIds }, "AuthService");
      return await httpClient.get(
        buildAuthUrl(API_ENDPOINTS.AUTH.GET_USER_NAME),
        { skipAuth: false, body: userIds }
      );
    } catch (error) {
      logger.error("Get user name failed", error, "AuthService");
      throw error;
    }
  },
  async login(email: string, password: string) {
    try {
      logger.info("Attempting user login", { email }, "AuthService");
      logger.debug(
        "Login request URL",
        { url: buildAuthUrl(API_ENDPOINTS.AUTH.LOGIN) },
        "AuthService"
      );

      return await httpClient.post(
        buildAuthUrl(API_ENDPOINTS.AUTH.LOGIN),
        { email, password },
        { skipAuth: true }
      );
    } catch (error) {
      logger.error("Login failed", error, "AuthService");
      logger.security("Login attempt failed", {
        email,
        error: error instanceof Error ? error.message : "Unknown error",
      });

      throw error;
    }
  },

  async logout() {
    try {
      logger.info("Attempting user logout", {}, "AuthService");
      logger.debug(
        "Logout request URL",
        { url: buildAuthUrl(API_ENDPOINTS.AUTH.LOGOUT) },
        "AuthService"
      );

      await httpClient.post(
        buildAuthUrl(API_ENDPOINTS.AUTH.LOGOUT),
        {},
        { skipAuth: false }
      );

      logger.info("Logout successful", "AuthService");
    } catch (error) {
      logger.error("Logout failed", error, "AuthService");
      throw error;
    }
  },

  async refreshToken() {
    try {
      logger.info("Refreshing token", {}, "AuthService");
      return await httpClient.refreshToken(
        buildAuthUrl(API_ENDPOINTS.AUTH.REFRESH),
        {
          skipAuth: true,
        }
      );
    } catch (error) {
      logger.error("Token refresh failed", error, "AuthService");
      throw error;
    }
  },

  async validateToken() {
    try {
      logger.debug("Validating token", {}, "AuthService");
      return await httpClient.get(buildAuthUrl(API_ENDPOINTS.AUTH.VALIDATE), {
        skipAuth: false,
      });
    } catch (error) {
      logger.error("Token validation failed", error, "AuthService");
      throw error;
    }
  },

  async changePassword(currentPassword: string, newPassword: string) {
    try {
      logger.info("Attempting password change", {}, "AuthService");
      logger.security("Password change attempt", { timestamp: Date.now() });

      return await httpClient.put(
        buildAuthUrl(API_ENDPOINTS.AUTH.CHANGE_PASSWORD),
        {
          currentPassword,
          newPassword,
        }
      );
    } catch (error) {
      logger.error("Password change failed", error, "AuthService");
      logger.security("Password change failed", {
        error: error instanceof Error ? error.message : "Unknown error",
      });

      throw error;
    }
  },

  async forgotPassword(email: string) {
    try {
      return await httpClient.post(
        buildAuthUrl(API_ENDPOINTS.AUTH.RESET_PASSWORD_REQUEST),
        {
          email,
        }
      );
    } catch (error) {
      logger.error("Forgot password failed:", error);
      throw error;
    }
  },

  async resetPassword(token: string, newPassword: string) {
    try {
      return await httpClient.post(
        buildAuthUrl(API_ENDPOINTS.AUTH.RESET_PASSWORD),
        {
          token,
          newPassword,
        }
      );
    } catch (error) {
      logger.error("Reset password failed:", error);
      throw error;
    }
  },

  async getUserInfo() {
    try {
      logger.info("Fetching user info", {}, "AuthService");
      return await httpClient.get(buildAuthUrl(API_ENDPOINTS.AUTH.USER_INFO), {
        skipAuth: false,
      });
    } catch (error) {
      logger.error("Failed to fetch user info", error, "AuthService");
      throw error;
    }
  },

  async updateUserInfo(updates: Partial<User>) {
    try {
      logger.info("Updating user info", updates, "AuthService");
      return await httpClient.put(
        buildAuthUrl(API_ENDPOINTS.AUTH.UPDATE),
        updates
      );
    } catch (error) {
      logger.error("Failed to update user info", error, "AuthService");
      throw error;
    }
  },

  async checkEmailExists(email: string) {
    try {
      logger.info("Checking if email exists", { email }, "AuthService");
      const endpoint = buildAuthUrl(API_ENDPOINTS.AUTH.CHECK_EMAIL, {
        email,
      });
      return await httpClient.get(endpoint, { skipAuth: true });
    } catch (error) {
      logger.error("Email check failed", error, "AuthService");
      throw error;
    }
  },

  async deleteAccount() {
    try {
      logger.info("Deleting user account", "AuthService");
      return await httpClient.put(
        buildAuthUrl(API_ENDPOINTS.AUTH.DELETE_REQUEST),
        { skipAuth: false }
      );
    } catch (error) {
      logger.error("Delete account failed", error, "AuthService");
      throw error;
    }
  },

  async register(user: Partial<User>) {
    try {
      logger.info("Registering new user", { email: user.email }, "AuthService");
      return await httpClient.post(
        buildAuthUrl(API_ENDPOINTS.AUTH.REGISTER),
        user,
        { skipAuth: true }
      );
    } catch (error) {
      logger.error("User registration failed", error, "AuthService");
      throw error;
    }
  },
};

File: mobile-app/travelmate/lib/features/Authentication/services/auth_service.dart
import 'package:travelmate/utils/helpers/env_file_reader.dart';
import 'package:travelmate/utils/http/http_client.dart';
import 'package:travelmate/utils/logging/logger.dart';

class AuthService {
  /// Map of function keys to endpoint paths
  static const Map<String, String> endpointMap = {
    'login': '/login',
    'register': '/register',
    'refresh': '/refresh',
    'checkEmailExists': '/check-email',
    'logout': '/logout',
    'requestPasswordReset': '/reset-password-request',
    'validateToken': '/validate',
    'getUserProfile': '/user-info',
    'updateUserProfile': '/update-user',
    'changePassword': '/change-password',
    'resendVerificationEmail': '/resend-verification',
    'deleteRequest': '/delete-request',
    'userNameByIds': ' /get-user-name',
  };

  /// Fetch Auth Endpoint from env file
  static Future<String> fetchAuthEndpoint() async {
    final endpoint = await TEnvFileReader.fetchAuthEndpoint();
    if (endpoint == null || endpoint.isEmpty) {
      throw Exception('Auth endpoint is not set in the environment file.');
    }
    return endpoint;
  }

  ///save User info
  static Future<void> saveUserInfo(dynamic userInfo) async {
    await TEnvFileReader.saveUserInfo(userInfo);
    TLoggerHelper.info('User info saved successfully.');
  }

  /// Fetch User info
  static Future<dynamic> fetchUserInfo() async {
    return await TEnvFileReader.fetchUserInfo();
  }

  /// Save token using TEnvFileReader
  static Future<void> saveTokenWithStorage(String token) async {
    await TEnvFileReader.saveUserToken(token);
    TLoggerHelper.info('Token saved successfully.');
  }

  /// Get stored token
  static Future<String?> getToken() async {
    return await TEnvFileReader.fetchUserToken();
  }

  /// Remove stored token
  static Future<void> removeToken() async {
    await TEnvFileReader.removeUserToken();
    TLoggerHelper.info('Token removed successfully.');
  }

  /// Save Refresh token
  static Future<void> saveRefreshToken(String refreshToken) async {
    await TEnvFileReader.saveRefreshToken(refreshToken);
    TLoggerHelper.info('Refresh Token saved successfully.');
  }

  /// Get stored Refresh token
  static Future<String?> getRefreshToken() async {
    return await TEnvFileReader.fetchRefreshToken();
  }

  /// Remove stored Refresh token
  static Future<void> removeRefreshToken() async {
    await TEnvFileReader.removeRefreshToken();
    TLoggerHelper.info('Refresh Token removed successfully.');
  }

  /// Check if user is logged in
  static Future<bool> isLoggedIn() async {
    final token = await getToken();
    final refreshToken = await getRefreshToken();
    TLoggerHelper.info('Refresh Token is present: ${refreshToken != null}');
    TLoggerHelper.info('User is logged in: ${token != null}');
    return (token != null && token.isNotEmpty) ||
        (refreshToken != null && refreshToken.isNotEmpty);
  }

  /// Login
  static Future<Map<String, dynamic>> login(
      String email, String password) async {
    TLoggerHelper.info('User login attempt with email: $email');
    try {
      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['login']!;
      final response = await THttpHelper.post(endpoint, {
        'email': email,
        'password': password,
      });

      final token = response['token'];
      final refreshToken = response['refresh_token'];
      final user = response['user'];

      if (token != null) {
        await saveTokenWithStorage(token);
        TLoggerHelper.info('Login successful, token saved.');
      } else {
        throw Exception('Token not found in response.');
      }

      if (refreshToken != null) {
        await saveRefreshToken(refreshToken);
        TLoggerHelper.info('Refresh token saved.');
      } else {
        TLoggerHelper.warning('Refresh token not found in response.');
      }

      if (user != null) {
        await saveUserInfo(user);
        TLoggerHelper.info('User info saved.');
      } else {
        TLoggerHelper.warning('User info not found in response.');
      }

      return {
        'success': true,
        'user': user,
        'token': token,
        'refreshToken': refreshToken,
      };
    } catch (e) {
      TLoggerHelper.error('Login failed: $e');
      rethrow;
    }
  }

  /// Register new user
  static Future<Map<String, dynamic>> register({
    required String name,
    required String email,
    required String password,
    required String dob,
    required String gender,
  }) async {
    TLoggerHelper.info('User registration attempt with email: $email');
    try {
      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['register']!;
      final response = await THttpHelper.post(endpoint, {
        'name': name,
        'email': email,
        'password': password,
        'dob': dob,
        'gender': gender,
      });

      final token = response['token'];
      final refreshToken = response['refresh_token'];
      final user = response['user'];

      if (token != null) {
        await saveTokenWithStorage(token);
        TLoggerHelper.info('Registration successful, token saved.');
      }

      if (refreshToken != null) {
        await saveRefreshToken(refreshToken);
        TLoggerHelper.info('Refresh token saved.');
      }

      return {
        'success': true,
        'user': user,
        'token': token,
        'refreshToken': refreshToken,
      };
    } catch (e) {
      TLoggerHelper.error('Registration failed: $e');
      rethrow;
    }
  }

  /// Refresh Token
  static Future<void> refreshToken() async {
    try {
      String? refreshToken = await getRefreshToken();
      if (refreshToken == null || refreshToken.isEmpty) {
        throw Exception('No refresh token found.');
      }
      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['refresh']!;
      final response =
          await THttpHelper.post(endpoint, null, token: refreshToken);
      final newToken = response['token'];
      final newRefreshToken = response['refresh_token'];
      if (newToken != null) {
        await saveTokenWithStorage(newToken);
        TLoggerHelper.info('Token refreshed successfully.');
      } else {
        throw Exception('New token not ound in response.');
      }
      if (newRefreshToken != null) {
        await saveRefreshToken(newRefreshToken);
        TLoggerHelper.info('Refresh token updated successfully.');
      } else {
        TLoggerHelper.warning('New refresh token not found in response.');
      }
    } catch (e) {
      TLoggerHelper.error('Token refresh failed: $e');
      rethrow;
    }
  }

  /// Check if email exists
  static Future<bool> checkEmailExists(String email) async {
    try {
      String authEndpoint = await fetchAuthEndpoint();
      final endpoint =
          '$authEndpoint${endpointMap['checkEmailExists']!}/$email';
      final response = await THttpHelper.post(endpoint, null);
      return response['exists'] ?? false;
    } catch (e) {
      TLoggerHelper.error('Email check failed: $e');
      return false;
    }
  }

  /// Logout
  static Future<void> logout() async {
    try {
      String? token = await getToken();
      if (token != null) {
        String authEndpoint = await fetchAuthEndpoint();
        final endpoint = authEndpoint + endpointMap['logout']!;
        await THttpHelper.post(endpoint, null, token: token);
      }

      await removeToken();
      await removeRefreshToken();
      TLoggerHelper.info('Logout successful');
    } catch (e) {
      TLoggerHelper.error('Logout failed: $e');
      // Still remove tokens even if API call fails
      await removeToken();
      await removeRefreshToken();
    }
  }

  /// Request Password Reset
  static Future<bool> requestPasswordReset(String email) async {
    try {
      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['requestPasswordReset']!;
      final response = await THttpHelper.post(endpoint, {
        'email': email,
      });

      TLoggerHelper.info('Password reset requested for: $email');
      return response['success'] ?? false;
    } catch (e) {
      TLoggerHelper.error('Password reset request failed: $e');
      return false;
    }
  }

  /// Validate Token
  static Future<bool> validateToken() async {
    try {
      String? token = await getToken();
      if (token == null) return false;

      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['validateToken']!;
      final response = await THttpHelper.post(endpoint, null, token: token);

      return response['valid'] ?? false;
    } catch (e) {
      TLoggerHelper.error('Token validation failed: $e');
      return false;
    }
  }

  /// Get User Profile
  static Future<Map<String, dynamic>?> getUserProfile() async {
    try {
      String? token = await getToken();
      if (token == null) return null;

      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['getUserProfile']!;
      final response = await THttpHelper.get(endpoint, token: token);
      return response['user'];
    } catch (e) {
      TLoggerHelper.error('Get user profile failed: $e');
      return null;
    }
  }

  /// Update User Profile
  static Future<Map<String, dynamic>?> updateUserProfile(
      Map<String, dynamic> profileData) async {
    try {
      String? token = await getToken();
      if (token == null) return null;

      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['updateUserProfile']!;
      final response =
          await THttpHelper.put(endpoint, profileData, token: token);

      TLoggerHelper.info('User profile updated successfully');
      return response['user'];
    } catch (e) {
      TLoggerHelper.error('Update user profile failed: $e');
      return null;
    }
  }

  /// Change Password
  static Future<bool> changePassword({
    required String oldPassword,
    required String newPassword,
  }) async {
    try {
      String? token = await getToken();
      if (token == null) return false;
      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['changePassword']!;
      final response = await THttpHelper.post(
          endpoint,
          {
            'old_password': oldPassword,
            'new_password': newPassword,
          },
          token: token);
      TLoggerHelper.info('Password change requested');
      return response['success'] ?? false;
    } catch (e) {
      TLoggerHelper.error('Change password failed: $e');
      return false;
    }
  }

  /// Resend Verification Email
  static Future<bool> resendVerificationEmail(String email) async {
    try {
      String authEndpoint = await fetchAuthEndpoint();
      String? token = await getToken();
      final endpoint = authEndpoint + endpointMap['resendVerificationEmail']!;
      final response = await THttpHelper.post(endpoint, null, token: token);
      TLoggerHelper.info('Verification email resend requested');
      return response['success'] ?? false;
    } catch (e) {
      TLoggerHelper.error('Resend verification email failed: $e');
      return false;
    }
  }

  /// Delete Account Request
  static Future<bool> deleteRequest() async {
    try {
      String? token = await getToken();
      if (token == null) return false;
      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['deleteRequest']!;
      final response = await THttpHelper.post(endpoint, null, token: token);
      TLoggerHelper.info('Delete account request sent');
      return response['success'] ?? false;
    } catch (e) {
      TLoggerHelper.error('Delete account request failed: $e');
      return false;
    }
  }

  /// Get User Names by IDs
  static Future<List<String>?> userNameByIds(List<String> userIds) async {
    try {
      String? token = await getToken();
      if (token == null) return null;
      String authEndpoint = await fetchAuthEndpoint();
      final endpoint = authEndpoint + endpointMap['userNameByIds']!;
      final response = await THttpHelper.post(
          endpoint,
          {
            'user_ids': userIds,
          },
          token: token);
      TLoggerHelper.info('Fetched user names by IDs');
      return (response['user_names'] as List<dynamic>?)
          ?.map((e) => e.toString())
          .toList();
    } catch (e) {
      TLoggerHelper.error('Get user names by IDs failed: $e');
      return null;
    }
  }
}
